ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -83,2575 +83,1531 @@
-
-	if (pc->output >= pc->start) {
-		if (c == (int)pc->needle.buffer[pc->needle_pos]) {
-			if (pc->needle_pos == 0) {
-				pc->found_pos = pc->output;			/* found position */
-			}
-			pc->needle_pos++;						/* needle pointer */
-			if (pc->needle_pos >= pc->needle_len) {
-				pc->matched_pos = pc->found_pos;	/* matched position */
-				pc->needle_pos--;
-				goto retry;
-			}
-		} else if (pc->needle_pos != 0) {
-retry:
-			h = (int *)pc->needle.buffer;
-			h++;
-			for (;;) {
-				pc->found_pos++;
-				p = h;
-				m = (int *)pc->needle.buffer;
-				n = pc->needle_pos - 1;
-				while (n > 0 && *p == *m) {
-					n--;
-					p++;
-					m++;
-				}
-				if (n <= 0) {
-					if (*m != c) {
-						pc->needle_pos = 0;
-					}
-					break;
-				} else {
-					h++;
-					pc->needle_pos--;
-				}
-			}
-		}
+	/* If the requested `from` and `length` would give us a range of bytes which starts
+	 * or ends in the middle of a multi-byte character, push the start position and/or
+	 * the end position forwards so we get only complete characters. This means the extracted
+	 * range may include slightly more or less bytes than requested.
+	 *
+	 * It's easiest when each character is encoded with a constant number of bytes: */
+	int bytes_per_char;
+	if (encoding->flag & (MBFL_ENCTYPE_WCS2BE | MBFL_ENCTYPE_WCS2LE)) {
+		bytes_per_char = 2;
+	} else if (encoding->flag & (MBFL_ENCTYPE_WCS4BE | MBFL_ENCTYPE_WCS4LE)) {
+		bytes_per_char = 4;
+	} else {
+		bytes_per_char = 1;
 	}
 
-	pc->output++;
-	return c;
+	from &= -bytes_per_char; /* push start position forward by zeroing low bits */
+	*length = MIN(*length, string->len - from) & -bytes_per_char;
+	*start = string->val + from;
 }
 
-static const unsigned char *mbfl_find_offset_utf8(
-		const unsigned char *str, const unsigned char *end, ssize_t offset) {
-	if (offset < 0) {
-		const unsigned char *pos = end;
-		while (offset < 0) {
-			if (pos <= str) {
-				return NULL;
-			}
+static inline void strcut_utf8(mbfl_string *string, size_t from, const unsigned char **start, size_t *length)
+{
+	/* It is easy to tell if we are in the middle of a multi-byte character
+	 * with UTF-8; all continuation bytes are like 10xxxxxx (in binary) */
+	unsigned char *search_pos = string->val + from;
+	while (((*search_pos & 0xC0) == 0x80) && (search_pos > string->val)) {
+		search_pos--;
+	}
+	*start = search_pos;
 
-			unsigned char c = *(--pos);
-			if (c < 0x80) {
-				++offset;
-			} else if ((c & 0xc0) != 0x80) {
-				++offset;
-			}
-		}
-		return pos;
+	size_t max_len = string->val + string->len - search_pos;
+	if (*length >= max_len) {
+		*length = max_len;
 	} else {
-		const unsigned char *u8_tbl = mbfl_encoding_utf8.mblen_table;
-		const unsigned char *pos = str;
-		while (offset-- > 0) {
-			if (pos >= end) {
-				return NULL;
-			}
-			pos += u8_tbl[*pos];
-		}
-		return pos;
+	  unsigned char *end = search_pos + *length;
+	  while (((*end & 0xC0) == 0x80) && (end > search_pos)) {
+			end--;
```

Comment:
Too much indentation

Reply:
Spaces snuck in there (instead of tabs). Don't know how that happened. Thanks for catching it.